get_primary_services/web_app.js

```diff
@@ -0,0 +1,171 @@
+      'writableAuxiliaries'];
+
+    verifyProperties(characteristic, trueProps, falseProps);
+  };
+
+  // Espruino does not set extended properties, so reliableWrite and
+  // writableAuxiliaries is always false.
+  const alwaysFalseProperties = ['reliableWrite', 'writableAuxiliaries'];
+
+  // Verify the properties of |secondCharacteristicUUID|, which should match
+  // those set in device_code.js.
+  const verifySecondCharacteristicProperties = (characteristic) => {
+    const trueProps = ['broadcast', 'write', 'writeWithoutResponse', 'indicate'];
+    const falseProps = ['read', 'notify',
+      'authenticatedSignedWrites'].concat(alwaysFalseProperties);
+    verifyProperties(characteristic, trueProps, falseProps);
+  };
+
+  // Get a string of all properties set to true.
+  const getCharacteristicProperties = (properties) => {
+    const propNames = ['read', 'notify', 'broadcast', 'writeWithoutResponse',
+      'write', 'indicate',
+      'authenticatedSignedWrites'].concat(alwaysFalseProperties);
+
+    let enabledProps = [];
+    propNames.forEach(propName => {
+      if (properties[propName]) {
+        enabledProps.push(propName);
+      }
+    });
+
+    if (enabledProps.length)
+      return enabledProps.join(', ');
+    return '<none>';
+  };
+
+  const logService = async (service) => {
+    let logText = [
+      ` Service UUID: ${service.uuid}, isPrimary: ${service.isPrimary}`
+    ];
+    if (service.device != remoteDevice)
+      throw 'Service device does not match connected device';
+    const characteristics = await service.getCharacteristics();
+    if (!characteristics.length) {
+      logText.push(`  No characteristics`);
+    } else {
+      characteristics.forEach(characteristic => {
+        logText.push(`  Characteristic UUID: ${characteristic.uuid}`);
+        logText.push(`   Props: ${getCharacteristicProperties(characteristic.properties)}`);
+        if (characteristic.uuid == firstCharacteristicUUID) {
+          verifyFirstCharacteristicProperties(characteristic);
+        } else if (characteristic.uuid == secondCharacteristicUUID) {
+          verifySecondCharacteristicProperties(characteristic);
+        }
+      });
+    }
+    logText.map(logInfo);
+  }
+
+  try {
+    const options = {
+      filters: [{ services: [getEspruinoPrimaryService()] }],
+      optionalServices: testServiceUUIDs
+    };
+    logInfo(`Requesting Bluetooth device with services ${JSON.stringify(testServiceUUIDs)}`);
+    remoteDevice = await navigator.bluetooth.requestDevice(options);
+    logInfo(`Device: ID: ${remoteDevice.id}, name: \"${remoteDevice.name}\"`),
+
+      remoteDevice.addEventListener('gattserverdisconnected', onGattDisconnected);
+    gattServer = await remoteDevice.gatt.connect();
+
+    logInfo(`Connected to GATT, requesting primary services...`);
+    const services = await gattServer.getPrimaryServices();
+
+    var actions = services.map(logService);
```

Comment:
Check that all the expected services appear.

Reply:
done